### PR TITLE
kube-addons: Use python container if python is not found on the machine.

### PR DESCRIPTION
**Pull request by yifan-gu**
Also add a Dockerfile to build a python image that has pyyaml.

Second try for #17241

### CODE REVIEW

**Comment by zmerlynn** [comment]
@k8s-bot test this

File: cluster/saltbase/salt/kube-addons/kube-addons.sh [hunk 1]
```diff
@@ -24,6 +24,18 @@ ADDON_CHECK_INTERVAL_SEC=${TEST_ADDON_CHECK_INTERVAL_SEC:-600}
 SYSTEM_NAMESPACE=kube-system
 token_dir=${TOKEN_DIR:-/srv/kubernetes}
 
+PYTHON=python
+
+function check_python() {
+  python --version > /dev/null 2>&1
+  if [[ $? != 0 ]]; then
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addons.sh]: nit: you can just elide this to `if ! python --version > /dev/null 2>&1; then`

File: cluster/saltbase/salt/kube-addons/kube-addons.sh [hunk 2]
```diff
@@ -24,6 +24,18 @@ ADDON_CHECK_INTERVAL_SEC=${TEST_ADDON_CHECK_INTERVAL_SEC:-600}
...
+    echo "No python on the machine, will build a python image"
+    PYTHON_IMAGE=python-slim-yaml
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addons.sh]: nit: `local -r PYTHON_IMAGE=python-slim-yaml`

File: cluster/saltbase/salt/kube-addons/kube-addons.sh [hunk 3]
```diff
@@ -24,6 +24,18 @@ ADDON_CHECK_INTERVAL_SEC=${TEST_ADDON_CHECK_INTERVAL_SEC:-600}
 SYSTEM_NAMESPACE=kube-system
 token_dir=${TOKEN_DIR:-/srv/kubernetes}
 
+PYTHON=python
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addons.sh]: Do you actually want to set this before `check_python` is called? If you leave it unbound, `check_python` will be required due to `nounset`. The way it's set right now, you don't have to call `check_python` at all.

Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addons.sh]: (As a corollary, it should probably be called `ensure_python`)

**Comment by zmerlynn** [comment]
This came out cleaner than I was expecting. Some minor nits, but generally looks fine.

**Comment by zmerlynn** [comment]
@k8s-bot test this

**Comment by zmerlynn** [comment]
Those @k8s-bot results are pretty clearly showing that this is failing to bring up add-ons on GCE, so something is amiss here.

**Comment by yifan-gu** [comment]
@zmerlynn I am taking a look

**Comment by zmerlynn** [comment]
LGTM! Thanks @yifan-gu!

**Comment by yifan-gu** [comment]
@zmerlynn hmm, I fails to use the docker images, hold on.

**Comment by yifan-gu** [comment]
@zmerlynn It keeps reporting `no available ip addresses on network` when executing `docker build`...

**Comment by yifan-gu** [comment]
@zmerlynn Turns out I cannot do docker build on the node because docker build will run to launch a container, and it's not able to get ip address.
I updated the PR to build the python image during `make` if `BUILD_PYTHON_IMAGE` env is set. And load the python image before running the addon script.

I tested, it works on non-python distros.

**Comment by yifan-gu** [comment]
The only thing I am not happy with is we need to keep the image name `python:2.7-slim-pyyaml` in both places (building script, and the addon script).

**Comment by yifan-gu** [comment]
@k8s-bot test this please

**Comment by yifan-gu** [comment]
@k8s-bot test this please

File: cluster/addons/README.md [hunk 4]
```diff
@@ -58,6 +58,8 @@ of:
      pods.
   1. Note that this cannot happen for Services as their version is always empty.
 
+Note that in order to run the updator script, python is required on the machine.
+For OS distros that doesn't have python installed, a python container will be used.
```
Comment (zmerlynn) [inline cluster/addons/README.md]: s/distros that doesn't/distros that don't/

File: cluster/addons/python-image/README.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# Python image
+
+The python image here is used by OS distros that doesn't have python installed to
```
Comment (zmerlynn) [inline cluster/addons/python-image/README.md]: s/distros that doesn't/distros that don't/

**Comment by zmerlynn** [comment]
LGTM except a couple of nits. I'm actually going to label it right now, since I'm going on vacation soon. If you want to fix the nits separately, go for it.

**Comment by yifan-gu** [comment]
@zmerlynn Thanks, typos fixed.

**Comment by yifan-gu** [comment]
From the log:

| FAIL    k8s.io/kubernetes/pkg/registry/generic/etcd     2.133s

Not seems to be related to this PR.

**Comment by yifan-gu** [comment]
@k8s-bot test this please

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Pull request merged**